# Analyzing borrowers’ risk of defaulting

Your project is to prepare a report for a bank’s loan division. You’ll need to find out if a customer’s marital status and number of children has an impact on whether they will default on a loan. The bank already has some data on customers’ credit worthiness.

Your report will be considered when building a **credit scoring** of a potential customer. A ** credit scoring ** is used to evaluate the ability of a potential borrower to repay their loan.

The project's goal:
*To find out the correlation between some given parameters of the customers (marital status, having children, income level) 
and their history in paying loans on time.
*from knowing the correlation and its strength we can recommend which parameters should be taken into account when formulating
potential customers' credit scoring.

## Open the data file and have a look at the general information. 

- Importing Pandas library

- Reading the file using read_csv 

- Getting dataframe general information

In [43]:
import pandas as pd

try:
    data_set = pd.read_csv('/datasets/credit_scoring_eng.csv')
except:    
    data_set = pd.read_csv('credit_scoring_eng.csv')
data_set.info()
data_set.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


children  days_employed  dob_years            education  education_id  \
0         1   -8437.673028         42    bachelor's degree             0   
1         1   -4024.803754         36  secondary education             1   
2         0   -5623.422610         33  Secondary Education             1   
3         3   -4124.747207         32  secondary education             1   
4         0  340266.072047         53  secondary education             1   
5         0    -926.185831         27    bachelor's degree             0   
6         0   -2879.202052         43    bachelor's degree             0   
7         0    -152.779569         50  SECONDARY EDUCATION             1   
8         2   -6929.865299         35    BACHELOR'S DEGREE             0   
9         0   -2188.756445         41  secondary education             1   

       family_status  family_status_id gender income_type  debt  total_income  \
0            married                 0      F    employee     0     40620.102   
1            married                 0      F    employee     0     17932.802   
2            married                 0      M    employee     0     23341.752   
3            married                 0      M    employee     0     42820.568   
4  civil partnership                 1      F     retiree     0     25378.572   
5  civil partnership                 1      M    business     0     40922.170   
6            married                 0      F    business     0     38484.156   
7            married                 0      M    employee     0     21731.829   
8  civil partnership                 1      F    employee     0     15337.093   
9            married                 0      M    employee     0     23108.150   

                               purpose  
0                purchase of the house  
1                         car purchase  
2                purchase of the house  
3              supplementary education  
4                    to have a wedding  
5                purchase of the house  
6                 housing transactions  
7                            education  
8                     having a wedding  
9  purchase of the house for my family

- There are 21,525 rows in the df. 

- There are 2,174 missing values in 'days employed' column and 2,174 in 'total income' column

### Conclusion

There are 21,525 rows in the file. 
There are missing values in 'days employed' and 'total income' columns. 
They both have the same number of mising values, so maybe these are the same rows. 
They both are quantitative data type, so we have to make some calulations in order to replace them. 

## Data preprocessing

### Processing missing values

- Checking if the missing values in 'days_employed' and 'total_income' are in the same rows
 - They both have the same number of missing values, so maybe there is a linkage.
 - They both quantitative data type, so we can't just replace them

- Creating the variable 'null_rows', which includes rows with both null columns
 - If the total number will not be higher then 2,174, that means they all the same rows

In [44]:
data_set.isnull().sum()

null_rows = data_set.loc[(data_set.days_employed.isnull()) & (data_set.total_income.isnull())]
null_rows

children  days_employed  dob_years            education  education_id  \
12            0            NaN         65  secondary education             1   
26            0            NaN         41  secondary education             1   
29            0            NaN         63  secondary education             1   
41            0            NaN         50  secondary education             1   
55            0            NaN         54  secondary education             1   
...         ...            ...        ...                  ...           ...   
21489         2            NaN         47  Secondary Education             1   
21495         1            NaN         50  secondary education             1   
21497         0            NaN         48    BACHELOR'S DEGREE             0   
21502         1            NaN         42  secondary education             1   
21510         2            NaN         28  secondary education             1   

           family_status  family_status_id gender    income_type  debt  \
12     civil partnership                 1      M        retiree     0   
26               married                 0      M  civil servant     0   
29             unmarried                 4      F        retiree     0   
41               married                 0      F  civil servant     0   
55     civil partnership                 1      F        retiree     1   
...                  ...               ...    ...            ...   ...   
21489            married                 0      M       business     0   
21495  civil partnership                 1      F       employee     0   
21497            married                 0      F       business     0   
21502            married                 0      F       employee     0   
21510            married                 0      F       employee     0   

       total_income                   purpose  
12              NaN         to have a wedding  
26              NaN                 education  
29              NaN    building a real estate  
41              NaN  second-hand car purchase  
55              NaN         to have a wedding  
...             ...                       ...  
21489           NaN         purchase of a car  
21495           NaN          wedding ceremony  
21497           NaN       building a property  
21502           NaN    building a real estate  
21510           NaN              car purchase  

[2174 rows x 12 columns]

In [45]:
len(null_rows)

2174

- There are exactly 2,174 rows in the filtered df. 
 - conclusion: where days_employed is null, also income_type is null. It make sense, if we take 'days_employed = 0' as 'not working'.

In [46]:
len(null_rows)/len(data_set)

0.10099883855981417

- That are 10% of the whole data

In [47]:
null_rows['income_type'].value_counts()

employee         1105
business          508
retiree           413
civil servant     147
entrepreneur        1
Name: income_type, dtype: int64

- I also checked what income types these people have, to check the hypothesis that they are all retired... and its wrong, they have all types of income

In [48]:
data_set['days_employed'].describe()

count     19351.000000
mean      63046.497661
std      140827.311974
min      -18388.949901
25%       -2747.423625
50%       -1203.369529
75%        -291.095954
max      401755.400475
Name: days_employed, dtype: float64

I tried to learn more about 'day_employed', by caculation some measures:
min = -18388
max = 401755
mean = 63046
median = -1203
negative numbers have no sense here. 
also the maximum value which implys for (when divided by 365) 1,100 working years...
mean and median are meaningless because of the negative values

I will try to see if these are just anomalies or whether they indicate that the whole column is corrupted
I will group the 'days_employed' to 3 groups: Minus, '0 to 60 years', and '60 years and more'
the '0 to 60 years' is the only logical group. lets see what is the distribution between the groups

In [49]:
def employed_groups(days):
       if days < 0:
            return 'Minus'
       if days < 22000:  
            return  '0 to 60 years' 
       return '60 years and more'         

data_set['emp_group'] = data_set['days_employed'].apply(employed_groups)
data_set['emp_group'].value_counts()

Minus                15906
60 years and more     5619
Name: emp_group, dtype: int64

There are 15,906 'Minus' and 5,619 '60 years and more'.
there are no rows with logical value of days_employed.
I want to learn about the distribution of each group.
The minus group:

In [50]:
data_set['days_employed'][data_set['emp_group'] == 'Minus'].min()
data_set['days_employed'][data_set['emp_group'] == 'Minus'].max()
data_set['days_employed'][data_set['emp_group'] == 'Minus'].mean()
data_set['days_employed'][data_set['emp_group'] == 'Minus'].median()

-1630.0193809778218

- minimum = -18388 (divided by 365 its about 50 years)
- maximum = -24 (about 1 month)
- mean = -2353 (~ 6.5 years)
- median = -1630 (~ 4.5 years)
- if we put them as absolute value the numbers are very sensible for employment period. so my assumption is that due to technical issue, the numbers are negative.

- I will replace the 'minus' values with their absolute value
 - now this group values make sense

In [51]:
data_set.loc[data_set['emp_group'] == 'Minus',['days_employed']] = data_set['days_employed'].abs()

- The '60 years and more' group:

In [52]:
data_set['days_employed'][data_set['emp_group'] == '60 years and more'].min()
data_set['days_employed'][data_set['emp_group'] == '60 years and more'].max()
data_set['days_employed'][data_set['emp_group'] == '60 years and more'].mean()
data_set['days_employed'][data_set['emp_group'] == '60 years and more'].median()

365213.3062657312

- minimum = 328,728 (divided by 8 ~50 years)
- maximum = 401,755 (divided by 8 ~ 1 month)
- mean = 365,004 (divided by 8 ~ 6.5 years)
- median = 365,213 (divided by 8 ~ 4.5 years)

- what are the income types of this group?

In [53]:
data_set['income_type'][data_set['emp_group'] == '60 years and more'].value_counts()

retiree          3856
employee         1105
business          508
civil servant     147
unemployed          2
entrepreneur        1
Name: income_type, dtype: int64

- Most of them are retiree. what rate?

In [54]:
'{:.2f}'.format(data_set['days_employed'].loc[(data_set.emp_group == '60 years and more') & (data_set.income_type == 'retiree')].count()/data_set['income_type'][data_set['emp_group'] == '60 years and more'].count())

'0.61'

 61% of the inordinate days_employed group are retiree
 I assumed their data represent someting else, not days. I tried to check if multiplying by 8 (hours a day) will make sense, but no.
 I will leave it as is for now.

- I will update the null values in 'days_employed' with the 'Minus' group mean

In [55]:
days_employed_mean = data_set['days_employed'][data_set['emp_group'] == 'Minus'].mean()
data_set['days_employed'].fillna(days_employed_mean, inplace = True)
data_set.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
 12  emp_group         21525 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


- days employed are set!
 - now I will check the 'total_income' distribution

In [56]:
data_set['total_income'].describe()

count     19351.000000
mean      26787.568355
std       16475.450632
min        3306.762000
25%       16488.504500
50%       23202.870000
75%       32549.611000
max      362496.645000
Name: total_income, dtype: float64

- Group the 'total_income' according to 'income_type' to see if there are significant groups:

In [57]:
data_set.groupby('income_type')['total_income'].mean()

income_type
business                       32386.793835
civil servant                  27343.729582
employee                       25820.841683
entrepreneur                   79866.103000
paternity / maternity leave     8612.661000
retiree                        21940.394503
student                        15712.260000
unemployed                     21014.360500
Name: total_income, dtype: float64

- Yes, grouping by 'income_type' gives us more information about the estimated income.
 - I will use the mean to fill the null values

- Define a function that recives a raw and updates 'total_income' null values with 'income_type' group's mean income
 - Apply the function "set_income" to the whole dataframe:

In [58]:
total_income_type = data_set.groupby('income_type')['total_income'].mean()

def set_income(raw):
    if pd.isnull(raw['total_income']):
        income = total_income_type[raw['income_type']]
        return income
    else:
        return raw['total_income']
  
data_set['total_income'] = data_set.apply(set_income, axis=1)

data_set['total_income'].describe()

count     21525.000000
mean      26783.346520
std       15665.112650
min        3306.762000
25%       17247.708000
50%       24309.013000
75%       32386.793835
max      362496.645000
Name: total_income, dtype: float64

### Conclusion

There were 2,174 rows with missing values in both 'days employed' column and 'total income' column.
1. Days employed:
I found out that the data of 'days employed' composed of two different groups, which both have unlogical data.
in one group the numbers are negetive, but their absolute value is ligical - I converted them to positive number.
The other group have very high values, which can't represent days of employment (the mean is 365,000 - 1,000 year of work...)
I left them as they are, maybe something will come up later on.
then I used the first group's mean days employed, to fill the null values.
2. Total income:
I found out there are significant differences in total income by the different income types, so Icalculated the mean total_income for each income_type and filled the null values with each group's mean income.
I chose the mean because for our purposes we will later group the income by income levels, so the differences between mean and median will not have much impact.

### Data type replacement

- Which columns' data type should be replaced?
 - days employed should be replaced into integer (days can get real number, but we are not loosing anything when using int)
 - same with total_income (we don't need the floating number information for our analysis)

In [59]:
data_set.info()

data_set['days_employed'] = data_set['days_employed'].astype('int')
data_set['total_income'] = data_set['total_income'].astype('int')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     21525 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      21525 non-null  float64
 11  purpose           21525 non-null  object 
 12  emp_group         21525 non-null  object 
dtypes: float64(2), int64(5), object(6)
memory usage: 2.1+ MB


### Conclusion

I replaced two columns data type from floating number to integer - days_employed and total_income. the reason is that in both of them the data will not be used for calculations, but for categorization.

### Processing duplicates

In [60]:
data_set.duplicated().sum()

54

- There are 54 duplicates (all the rows' values). thats only 0.2% of the data
 - And since there is no unique id for each row, I can't be sure that these rows are duplicates. they might be original rows.

2.6 Conclusion

Since the duplicated rows can be real 2 similar records, and since they consist just 0.2% of the data, I decided not to drop them

### Categorizing Data

- education and education id
- family status
 - These two values has both the id & description in the table, which takes a lot of space and add 'noise',
- I will divide the table into 1 main table and 2 Id tables. It will be more orgenized to create one ID table for the whole categories, and another master table for dhe categories key, But for 2 categories only, its not worth it.

- Main customer details df:

In [61]:
customer_details = data_set[['children','days_employed','dob_years','education_id','family_status_id','gender','income_type','debt','total_income','purpose','emp_group']]
customer_details

children  days_employed  dob_years  education_id  family_status_id  \
0             1           8437         42             0                 0   
1             1           4024         36             1                 0   
2             0           5623         33             1                 0   
3             3           4124         32             1                 0   
4             0         340266         53             1                 1   
...         ...            ...        ...           ...               ...   
21520         1           4529         43             1                 1   
21521         0         343937         67             1                 0   
21522         1           2113         38             1                 1   
21523         3           3112         38             1                 0   
21524         2           1984         40             1                 0   

      gender income_type  debt  total_income                  purpose  \
0          F    employee     0         40620    purchase of the house   
1          F    employee     0         17932             car purchase   
2          M    employee     0         23341    purchase of the house   
3          M    employee     0         42820  supplementary education   
4          F     retiree     0         25378        to have a wedding   
...      ...         ...   ...           ...                      ...   
21520      F    business     0         35966     housing transactions   
21521      F     retiree     0         24959        purchase of a car   
21522      M    employee     1         14347                 property   
21523      M    employee     1         39054        buying my own car   
21524      F    employee     0         13127             to buy a car   

               emp_group  
0                  Minus  
1                  Minus  
2                  Minus  
3                  Minus  
4      60 years and more  
...                  ...  
21520              Minus  
21521  60 years and more  
21522              Minus  
21523              Minus  
21524              Minus  

[21525 rows x 11 columns]

- convert the education string into lowercase 
- creating education values table
- deleting the duplicated rows

In [62]:
data_set['education'] = data_set['education'].str.lower()

edu_id_table = data_set[['education_id',('education')]]

edu_id_table = edu_id_table.drop_duplicates().reset_index(drop=True)
edu_id_table

education_id            education
0             0    bachelor's degree
1             1  secondary education
2             2         some college
3             3    primary education
4             4      graduate degree

- And do the same for the family status 
 - creating family_status values table
 - deleting the duplicated rows

In [63]:
data_set['family_status'] = data_set['family_status'].str.lower()

fs_id_table = data_set[['family_status_id',('family_status')]]

fs_id_table = fs_id_table.drop_duplicates().reset_index(drop=True)
fs_id_table

family_status_id      family_status
0                 0            married
1                 1  civil partnership
2                 2    widow / widower
3                 3           divorced
4                 4          unmarried

- lets exemine the children data

In [64]:
customer_details['children'].value_counts()

 0     14149
 1      4818
 2      2055
 3       330
 20       76
-1        47
 4        41
 5         9
Name: children, dtype: int64

- There are 47 rows with -1 children and 76 rows with 20 children (Total 123 rows)
 - They consist 0.6% of the population
 - I want to exemine this inspections and see if I can get with some hypothesis

In [65]:
twenty_kids = customer_details[customer_details['children'] == 20]
minus_kids = customer_details[customer_details['children'] == -1]

twenty_kids.head(10)
minus_kids.head(10)

children  days_employed  dob_years  education_id  family_status_id  \
291         -1           4417         46             1                 1   
705         -1            902         50             1                 0   
742         -1           3174         57             1                 0   
800         -1         349987         54             1                 4   
941         -1           2353         57             1                 0   
1363        -1           1195         55             1                 0   
1929        -1           1461         38             1                 4   
2073        -1           2539         42             1                 3   
3814        -1           3045         26             1                 1   
4201        -1            901         41             1                 0   

     gender    income_type  debt  total_income  \
291       F       employee     0         16450   
705       F  civil servant     0         22061   
742       F       employee     0         10282   
800       F        retiree     0         13806   
941       F        retiree     0         21940   
1363      F       business     0         11128   
1929      M       employee     0         17459   
2073      F       business     0         26022   
3814      F  civil servant     0         21102   
4201      F  civil servant     0         36220   

                               purpose          emp_group  
291                  profile education              Minus  
705                       car purchase              Minus  
742            supplementary education              Minus  
800            supplementary education  60 years and more  
941                  buying my own car  60 years and more  
1363                 profile education              Minus  
1929             purchase of the house              Minus  
2073             purchase of the house              Minus  
3814                  having a wedding              Minus  
4201  transactions with my real estate              Minus

- No, I can't find any correlation to anything that could explain the outre values. 
 - I assume their source is some corruption or problem with data entry
 - Since these outre values consist only 0.6% of the population, I believe I can drop them

In [66]:
customer_details = customer_details.loc[(customer_details['children'] >= 0) & (customer_details['children'] < 20)]

customer_details['children'].value_counts()

0    14149
1     4818
2     2055
3      330
4       41
5        9
Name: children, dtype: int64

- In order to evaluate the risk in relation to having kids, I will group the data by having/not having 

- I will do it by replacing the children column number value by Yes(have children)/No(0 children)

In [67]:
import sys
import warnings
if not sys.warnoptions:
       warnings.simplefilter("ignore")

In [68]:
customer_details.loc[customer_details['children'] != 0,['children']] = 'Yes'
customer_details.loc[customer_details['children'] == 0,['children']] = 'No'
customer_details

children  days_employed  dob_years  education_id  family_status_id  \
0          Yes           8437         42             0                 0   
1          Yes           4024         36             1                 0   
2           No           5623         33             1                 0   
3          Yes           4124         32             1                 0   
4           No         340266         53             1                 1   
...        ...            ...        ...           ...               ...   
21520      Yes           4529         43             1                 1   
21521       No         343937         67             1                 0   
21522      Yes           2113         38             1                 1   
21523      Yes           3112         38             1                 0   
21524      Yes           1984         40             1                 0   

      gender income_type  debt  total_income                  purpose  \
0          F    employee     0         40620    purchase of the house   
1          F    employee     0         17932             car purchase   
2          M    employee     0         23341    purchase of the house   
3          M    employee     0         42820  supplementary education   
4          F     retiree     0         25378        to have a wedding   
...      ...         ...   ...           ...                      ...   
21520      F    business     0         35966     housing transactions   
21521      F     retiree     0         24959        purchase of a car   
21522      M    employee     1         14347                 property   
21523      M    employee     1         39054        buying my own car   
21524      F    employee     0         13127             to buy a car   

               emp_group  
0                  Minus  
1                  Minus  
2                  Minus  
3                  Minus  
4      60 years and more  
...                  ...  
21520              Minus  
21521  60 years and more  
21522              Minus  
21523              Minus  
21524              Minus  

[21402 rows x 11 columns]

- Categorizing total_income by income levels
 - I will devide the incomes to four groups: first 25%,25%-50%,50%-75%, 75% and above
 - Define functions that gets the total_income and returns income level
 - Apply the function to the dataframe on total_income column, to create income level column

In [69]:
customer_details['total_income'].describe()

def income_level_grouping(income):
       if income <= 17200:
            return 'less then 17200'
       if income <= 24300:  
            return  '17200-24300'
       if income <= 32400:
            return '24300-32400'
       return  '32400 and above'

customer_details['income_level'] = customer_details['total_income'].apply(income_level_grouping)
customer_details['income_level'].value_counts()

24300-32400        5834
17200-24300        5356
less then 17200    5333
32400 and above    4879
Name: income_level, dtype: int64

Categorize loan purpose:

- Define function that gets the loan purpose as was reported and returns the right purpose category
- Apply the function to all the rows creates a new column 'purpose category'   

In [70]:
import nltk
from nltk.stem import SnowballStemmer
from nltk.stem import WordNetLemmatizer
english_stemmer = SnowballStemmer('english')
wordnet_lemma = WordNetLemmatizer()
purpose_step_dict = {'wedding':['wed'],'real_estate':['hous','properti','estat'],'car':['car'],'education':['educ','univers']}

def categorize(purpose_text):
        loan_purpose = nltk.word_tokenize(purpose_text.lower())

        for word in loan_purpose:
            word = english_stemmer.stem(word)
            if word in purpose_step_dict.get('wedding'):
                return 'wedding'
                break
            if  word in purpose_step_dict.get('real_estate'):
                return 'real estate'
                break
            if word in purpose_step_dict.get('car'):
                return 'car'
                break
            if word in purpose_step_dict.get('education'):
                return 'education'
                break
            else:
                continue
                        
customer_details.loc[:,'purpose_category'] = customer_details.loc[:,'purpose'].apply(categorize)
customer_details['purpose_category'].value_counts()

real estate    10780
car             4288
education       3997
wedding         2337
Name: purpose_category, dtype: int64

### Conclusion

I had a lot of work in categorizing the data. I did it in order to build the basis for later analysis by categories:
1. I took out the educationdescription and family_status description, created dedicated id_description dataframe, and left onlt the id's in the main dataframe
2. Categorized the children data into two categories - having children Yes/No
3. Categorized total income into 4 income levels according to their percentile.
4. Categorized loan purpose free text that actualy holds 4 different purposes, into 4 purpose types. did it by tokenizing the text and stemming each word. using functions that checks if the word belongs to one of the lists in the dictionary in built for it.

## Answer these questions

- Is there a relation between having kids and repaying a loan on time?

- First, I would like to calculate the mean debt_rate for the whole population:

In [71]:
mean_debt_rate = customer_details['debt'].sum()/customer_details['debt'].count()*100
mean_debt_rate.round(2)

8.09

- The mean debt rate (the chance that a customer has a history of defaulting a loan) is 8.09%

 - Define pivot table for the index children Yes/No
 - Calculating the rate of people having defaulting on a loan for each category 

In [72]:
children_data_pivot = pd.pivot_table(customer_details,index=['children'],values = ['debt'],aggfunc=[len,sum])
             
children_data_pivot['debt_rate'] = children_data_pivot['sum']/children_data_pivot['len']*100
children_data_pivot.sort_values(by = 'debt_rate', ascending = False)

len   sum debt_rate
           debt  debt          
children                       
Yes        7253   669  9.223769
No        14149  1063  7.512898

### Conclusion

Yes, There is a relation betwwen having kids and repaying loans on time: we can see that 9.2% of people with kids have history of defaulting paying loans, which is above the population mean, and only 7.5% of the people without kids, which is below the 8.09% mean.

- Is there a relation between marital status and repaying a loan on time?

 - Creating pivot table for debt history by family_status categories
 - Calculating the rate of people having defaulting on a loan for each category (# of children)
 - Adding family status description and creating new df
 - Updating columns names
 - Sorting the data by loan-payment-defaulting rate

In [73]:
family_data_pivot = pd.pivot_table(customer_details,index=['family_status_id'],values = ['debt'],aggfunc=[len,sum])
               
family_data_pivot['debt_rate'] = family_data_pivot['sum']/family_data_pivot['len']*100

debt_rate_by_family_status = family_data_pivot.merge(fs_id_table,on = 'family_status_id', how = 'left')

In [74]:
debt_rate_by_family_status.set_axis(['family_status_id', 'Total_customers', 'Total_debt', 'debt_rate','family_status'], axis='columns',inplace = True)

In [75]:
debt_rate_by_family_status.sort_values(by = 'debt_rate', ascending = False)

family_status_id  Total_customers  Total_debt  debt_rate      family_status
4                 4             2799         273   9.753483          unmarried
1                 1             4160         385   9.254808  civil partnership
0                 0            12302         927   7.535360            married
3                 3             1189          84   7.064760           divorced
2                 2              952          63   6.617647    widow / widower

### Conclusion

Yes, There is a relation betwwen marital status and repaying loans on time: we can see that unmarried people ot people in civil partnership have higher then 9.25% unrepayed loans rate, which is more than the everage, while married or has-been-married people have 7.5% rate and below.

- Is there a relation between income level and repaying a loan on time?

- Create pivot table for debt(loan repayment history) by income level 
 - Calculating the rate of bad history for each level 
 - Sort the rows by debt rate 

In [76]:
income_data_pivot = pd.pivot_table(customer_details,index=['income_level'],values = ['debt'],aggfunc=[len,sum])

income_data_pivot['debt_rate'] = income_data_pivot['sum']/income_data_pivot['len']*100

income_data_pivot.sort_values(by = 'debt_rate', ascending = False)

len  sum debt_rate
                 debt debt          
income_level                        
17200-24300      5356  470  8.775205
24300-32400      5834  492  8.433322
less then 17200  5333  427  8.006750
32400 and above  4879  343  7.030129

### Conclusion

There is no relation between income level and repaying a loan on time in the 3 lower levels - they are all somewhere around the average 8.09% of bad history. But - high level of income (the top 25%) has correlation with repaying a loan on time! this group has only 7.03% of bad history

- How do different loan purposes affect on-time repayment of the loan?

 - Create pivot table for debt(loan repayment history) by purpose categories
 - Calculating the rate of bad history for each category
 - Sort the rows by debt rate 

In [77]:
purpose_data_pivot = pd.pivot_table(customer_details,index=['purpose_category'],values = ['debt'],aggfunc=[len,sum])

purpose_data_pivot['debt_rate'] = purpose_data_pivot['sum']/purpose_data_pivot['len']*100

purpose_data_pivot.sort_values(by = 'debt_rate', ascending = False)

len  sum debt_rate
                   debt debt          
purpose_category                      
car                4288  400  9.328358
education          3997  369  9.231924
wedding            2337  183  7.830552
real estate       10780  780  7.235622

### Conclusion

Loans taken for the purpose of buying a car or for education, have higher rates of default repayment (more then 9.23%, while the evarage is 8.09%) then loans taken for wedding celebration or for buying real estate (7.83% and below). 

## General conclusion

1)Data Preperation:
Missing values- The data file had missing values in two columns - 'days_employed' and 'total_income'.
a.Days_employed: before fixing the null values I had to take care of the anomalies in the data. there were two types of 'days_employed': group of (-) negative values, and group of non-logic high values. what I did is to convert the negative values to their absolute value. After that, I used their mean to update the null values. I left the superhigh values as they are.
(I am not sure I had to take care of this column at all, because the 'days_employed' column don't use for any analysis here)
b.Total_income: I found there are significant differences between the mean/meadian of total_income by groups of the income_type. so I filled the null rows with the group's total income median.

Duplicates - There were 54 duplicates but I decided not to remove them because there are good chances they are real.

Anomalies - I found anomalies in the childrens number values. there were some (-1) and (20) values, which are not logic. since they form only 0.6% of the data, I decided to drop them.

Categorizing and findings:
I Found out some interesting information regarding customers potential credit score:
1. After categorizing the data into two groups of having/not having children I found that:                                        Not having kids is related to better chances of repaying a loan on time
2. After categorizing the data by family status into two groups of married or been married/never been married I foune that:
   Being married in the present or in the past is related to better chances of repaying a loan on time
3. After categorizing the data to four groups by the total income quortiles I found that:
   After grouping the loan purpose into the Income level in the high 25% percentage is related to better chances of repaying a loan on time
4. After I found the root purpose from the loan_purpose text and grouped it into four purposes I found that:
   Loan purpose of buying a property or paying a wedding are related to better chances of repaying a loan on time